extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/tracing/cdi/WithSpanInterceptor.java

```diff
@@ -53,35 +57,100 @@ public Object span(final ArcInvocationContext invocationContext) throws Exceptio
                 invocationContext.getParameters(),
                 invocationContext.getInterceptorBindings());
 
+        final Class<?> returnType = invocationContext.getMethod().getReturnType();
         Context parentContext = Context.current();
-        Context spanContext = null;
-        Scope scope = null;
         boolean shouldStart = instrumenter.shouldStart(parentContext, methodRequest);
-        if (shouldStart) {
-            spanContext = instrumenter.start(parentContext, methodRequest);
-            scope = spanContext.makeCurrent();
-        }
 
-        try {
-            Object result = invocationContext.proceed();
+        if (!shouldStart) {
+            return invocationContext.proceed();
+        }
 
-            if (shouldStart) {
-                instrumenter.end(spanContext, methodRequest, null, null);
+        if (isUni(returnType)) {
+            final Context currentSpanContext = instrumenter.start(parentContext, methodRequest);
+            final Scope currentScope = currentSpanContext.makeCurrent();
+            try {
+                return ((Uni<Object>) invocationContext.proceed()).onTermination().invoke((o, throwable, aCancelled) -> {
+                    if (aCancelled) {
+                        instrumenter.end(currentSpanContext, methodRequest, null, new RuntimeException("Cancelled"));
```

Comment:
Not sure it makes sense, but the JDK has a cancellation exception.

Reply:
will try

Reply:
done